### ISSUE TITLE
[YSQL] prevent inconsistent state on DROP INDEX

### ISSUE THREAD

**Issue by jaki** [yugabyte/yugabyte-db#4360]
It is possible that, when dropping an index, the DocDB index is dropped (`YBCDropIndex`) but the postgres index is not dropped (`index_drop`).  This can happen if the `DROP INDEX` is interrupted, such as with ctrl+c.  This causes postgres and DocDB to differ.

Apply this small patch to make it much easier to properly time the interrupt:

```diff
diff --git i/src/postgres/src/backend/catalog/dependency.c w/src/postgres/src/backend/catalog/dependency.c
index 9ebe04454..80f0a7d60 100644
--- i/src/postgres/src/backend/catalog/dependency.c
+++ w/src/postgres/src/backend/catalog/dependency.c
@@ -1139,6 +1139,9 @@ doDeletion(const ObjectAddress *object, int flags)
 
 						RelationClose(index);
 					}
+					elog(LOG, "sleeping");
+					pg_usleep(60 * 1000 * 1000);
+					elog(LOG, "sleep done");
 					index_drop(object->objectId, concurrent);
 				}
 				else
```

Then, run

```sql
CREATE TABLE a (i int);
CREATE INDEX ON a (i);
DROP INDEX i_a_idx;
```

and ctrl+c during the drop index while it is sleeping.  Then, run

```sql
SELECT i FROM a WHERE i = 1;
````

to trigger an index scan.  It should fail with

```
ERROR:  Not found: Error loading table with oid 16387 in database with oid 12457: The object '000030a9000030008000000000000000.a_i_idx' does not exist
```

I somewhat expected the limited transactional DDL commit b68b84bfd46b6a771d828c66477a477aa0f71844 to cover this, but I guess it's not quite the same because this is user table data being deleted, not system table modifications.

**Comment by ndeodhar** [comment]
The same problem can happen with `DROP TABLE` too and we should handle that as well in the solution.

Per prior discussion with @m-iancu :

To fix this inconsistent state, we should drop the docdb table (or index) *after* the postgres transaction is committed.
Note that there will still be a window where the postgres transaction commits and drops the index, but the docdb index is not dropped. Such a situation will be solved once we have online schema migration framework and docdb removes entries related to dropped postgres entries (similar to how we handle namespace creation failure, for example). In the interim, users can use `yb-admin delete_index` or `yb-admin delete_table` to drop the docdb table/index.

Note that this is still an improvement over the current state where YSQL metadata is inconsistent. With the proposed solution to drop table/index after postgres transaction is committed, we'll have an orphaned docdb table but that won't cause any issues for user queries.

**Comment by m-iancu** [comment]
FIxed by 3253827.
